homeassistant/components/media_player/bluesound.py

```diff
@@ -0,0 +1,812 @@
+        self._services_items = []
+        self._preset_items = []
+        self._sync_status = {}
+        self._status = None
+        self._last_status_update = None
+        self._is_online = False
+        self._retry_remove = None
+        self._lastvol = None
+        self._init_callback = init_callback
+        if self._port is None:
+            self._port = DEFAULT_PORT
+
+# Internal methods
+    @staticmethod
+    def _try_get_index(string, seach_string):
+        try:
+            return string.index(seach_string)
+        except ValueError:
+            return -1
+
+    @asyncio.coroutine
+    def _internal_update_sync_status(self, on_updated_cb=None,
+                                     raise_timeout=False):
+        resp = None
+        try:
+            resp = yield from self.send_bluesound_command(
+                'SyncStatus',
+                raise_timeout, raise_timeout)
+        except:
+            raise
+
+        if not resp:
+            return None
+        self._sync_status = resp['SyncStatus'].copy()
+
+        if not self._name:
+            self._name = self._sync_status.get('@name', self.host)
+        if not self._brand:
+            self._brand = self._sync_status.get('@brand', self.host)
+        if not self._model:
+            self._model = self._sync_status.get('@model', self.host)
+        if not self._icon:
+            self._icon = self._sync_status.get('@icon', self.host)
+        if not self._model_name:
+            self._model_name = self._sync_status.get('@modelName', self.host)
+
+        if on_updated_cb:
+            on_updated_cb()
+        return True
+# END Internal methods
+
+# Poll functionality
+    @asyncio.coroutine
+    def _start_poll_command(self):
+        """"Loop which polls the status of the player."""
+        try:
+            while True:
+                yield from self.async_update_status()
+
+        except (asyncio.TimeoutError, ClientError):
+            _LOGGER.info("Bluesound node %s is offline, retrying later",
+                         self._name)
+            yield from asyncio.sleep(NODE_OFFLINE_CHECK_TIMEOUT,
+                                     loop=self._hass.loop)
+            self.start_polling()
+
+        except CancelledError:
+            _LOGGER.debug("Stopping bluesound polling of node %s", self._name)
+        except:
+            _LOGGER.exception("Unexpected error in %s", self._name)
+            raise
+
+    def start_polling(self):
+        """Start the polling task."""
+        self._polling_task = self._hass.async_add_job(
+            self._start_poll_command())
+
+    def stop_polling(self):
+        """Stop the polling task."""
+        self._polling_task.cancel()
+# END Poll functionality
+
+# Initiator
+    @asyncio.coroutine
+    def async_init(self, what=None):
+        """Initiate the player async."""
+        if what is None:
+            # if it's the first request.
+            # let's make the request in a new thread
```

Comment:
Hi, finally I have been looking into this.

I can't get it to be a non blocking thread when using hass.async_add_job. The call will be blocking the startup of HA. Any ideas on how to solve this? I don't want this component to block the startup.

This is the log when using **hass.async_add_job**

```
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.setup] Setting up media_player
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.setup] Setting up discovery
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.setup] Setup of domain discovery took 0.0 seconds.
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=discovery>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.components.media_player] Setting up media_player.bluesoundotepa
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.components.media_player.bluesound] calling URL: http://192.168.1.132:11000/SyncStatus
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=turn_on>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=turn_off>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=toggle>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_up>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_down>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_play_pause>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_play>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_pause>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_stop>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_next_track>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_previous_track>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=clear_playlist>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_set>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_mute>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_seek>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=select_source>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=play_media>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=shuffle_set>
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.setup] Setup of domain media_player took 0.2 seconds.
2017-07-24 16:57:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=media_player>
2017-07-24 16:57:51 INFO (MainThread) [homeassistant.components.media_player.bluesound] Timeout with Bluesound: 192.168.1.132
2017-07-24 16:57:51 INFO (MainThread) [homeassistant.components.media_player.bluesound] Bluesound node 192.168.1.132 is offline, retrying later
2017-07-24 16:57:51 INFO (MainThread) [homeassistant.bootstrap] Home Assistant initialized in 20.72s
```

And this is the log when using **async_track_time_interval**
```
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.setup] Setting up media_player
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.setup] Setting up discovery
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.setup] Setup of domain discovery took 0.0 seconds.
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=discovery>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.components.media_player] Setting up media_player.bluesound
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=turn_on>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=turn_off>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=toggle>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_up>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_down>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_play_pause>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_play>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_pause>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_stop>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_next_track>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_previous_track>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=clear_playlist>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_set>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=volume_mute>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=media_seek>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=select_source>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=play_media>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=media_player, service=shuffle_set>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.setup] Setup of domain media_player took 0.2 seconds.
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=media_player>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.bootstrap] Home Assistant initialized in 0.81s
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Starting Home Assistant core loop
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Starting Home Assistant
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Bus:Handling <Event homeassistant_start[L]>
2017-07-24 16:58:53 INFO (MainThread) [homeassistant.core] Timer:starting
2017-07-24 16:58:54 INFO (MainThread) [homeassistant.components.http] Serving / to 10.225.70.119 (auth: True)
2017-07-24 16:58:54 INFO (MainThread) [homeassistant.components.http] Serving /api/websocket to 10.225.70.119 (auth: True)
2017-07-24 16:58:54 INFO (MainThread) [homeassistant.components.http] Serving /api/websocket to 10.225.70.119 (auth: True)
2017-07-24 16:58:54 INFO (MainThread) [homeassistant.components.http] Serving /api/themes to 10.225.70.119 (auth: True)
2017-07-24 16:58:54 INFO (MainThread) [homeassistant.components.media_player.bluesound] calling URL: http://192.168.1.132:11000/SyncStatus
```